### PR TITLE
new zombie: necro-boomer

### PR DESCRIPTION
**Pull request by KorGgenT**
#### Summary

```SUMMARY: Content "new zombie: necro-boomer"```
#### Purpose of change
new scarier monsters for larger endgame

#### Describe the solution
added a new boomer type zombie: the necro-boomer. on death, it revives all zombies and then upgrades all zombies within a radius of 10. aded it to the UPGRADE monstergroup and rarely to the FAT monstergroup.

#### Describe alternatives you've considered
eh.

#### Testing
spawned a group of zombies, killed them, and then spawned a few more, and then killed a necroboomer. even the revived zombies upgrade. (which is what i wanted)

#### Additional context
I kinda don't like how the death function code is structured; i'd really like to make it possible to use spells for monster death functions for more json ability, but at the time i just wanted to make a new zombo.

### CODE REVIEW

File: data/json/monsters/zed_misc.json [hunk 1]
```diff
@@ -707,6 +707,15 @@
       "FILTHY"
     ]
   },
+  {
+    "id": "mon_zombie_necro_boomer",
+    "type": "MONSTER",
+    "name": { "str": "zombie necro-boomer" },
+    "description": "This zombie looks like the zombie necromancer, but it blisters with oozing black pustules and boats to an enormous size.  It seems ready to pop.",
```
Comment (KorGgenT) [inline data/json/monsters/zed_misc.json:714]:
```suggestion
    "description": "This zombie looks like the zombie necromancer, but it blisters with oozing black pustules and bloats to an enormous size.  It seems ready to pop.",
```

**Comment by procrustesmethod** [comment]
"At first this creature looks like nothing more than a grotesque and oleaginous husk, bloated and punctured by jet-black pustules. When approached its glowing red eyes and an aching grin take form; its skin tears and guts teem with unmatched fecundity, as its gaze inspires fear of nothing less than cosmic, inhuman ecstasy. "

^ Proposed new description of zombie necro-boomer.

**Comment by ABeligerentMoron** [comment]
Goddamn that's just straight up evil

File: data/json/monsters/zed_misc.json [hunk 2]
```diff
@@ -707,6 +707,15 @@
       "FILTHY"
     ]
   },
+  {
+    "id": "mon_zombie_necro_boomer",
+    "type": "MONSTER",
+    "name": { "str": "zombie necro-boomer" },
+    "description": "This zombie looks like the zombie necromancer, but it blisters with oozing black pustules and bloats to an enormous size.  It seems ready to pop.",
```
Comment (KorGgenT) [inline data/json/monsters/zed_misc.json:714]:
```suggestion
    "description": "At first this creature looks like nothing more than a grotesque and oleaginous husk, bloated and punctured by jet-black pustules. When approached its glowing red eyes and an aching grin take form; its skin tears and guts teem with unmatched fecundity, as its gaze inspires fear of nothing less than cosmic, inhuman ecstasy.",
```

**Comment by Night-Pryanik** [comment]
I don't like this. Zombie necromancer and zombie master were on the edge with their magic abilities, and this monster steps far over the border. It's just a combined version of two previous monsters with no new abilities, no distinctive features, and, more importantly, no justification. It clearly smells of "difficulty for the sake of difficulty".

File: data/json/monsters/zed_misc.json [hunk 3]
```diff
@@ -707,6 +707,15 @@
       "FILTHY"
     ]
   },
+  {
+    "id": "mon_zombie_necro_boomer",
+    "type": "MONSTER",
+    "name": { "str": "zombie necro-boomer" },
+    "description": "At first this creature looks like nothing more than a grotesque and oleaginous husk, bloated and punctured by jet-black pustules. When approached its glowing red eyes and an aching grin take form; its skin tears and guts teem with unmatched fecundity, as its gaze inspires fear of nothing less than cosmic, inhuman ecstasy.",
```
Comment (Night-Pryanik) [inline data/json/monsters/zed_misc.json:714]:
```suggestion
    "description": "At first this creature looks like nothing more than a grotesque and oleaginous husk, bloated and punctured by jet-black pustules.  When approached its glowing red eyes and an aching grin take form; its skin tears and guts teem with unmatched fecundity, as its gaze inspires fear of nothing less than cosmic, inhuman ecstasy.",
```

File: src/mondeath.cpp [hunk 4]
```diff
@@ -899,3 +899,26 @@ void mdeath::conflagration( monster &z )
     sounds::sound( z.pos(), 24, sounds::sound_t::combat, explode, false, "explosion", "small" );
 
 }
+
+void mdeath::necro_boomer( monster &z )
+{
+    std::string explode = string_format( _( "a %s explode!" ), z.name() );
```
Comment (Night-Pryanik) [inline src/mondeath.cpp:905]:
```suggestion
    std::string explode = string_format( _( "a %s explodes!" ), z.name() );
```

**Comment by Amoebka** [comment]
This feels rather silly. Necromancers and masters are edgy magicians that seemingly use "spells" (gesticulate wildly) to do what they do, while this one is a fatty that does magic by exploding into a pool of gore. Kinda takes away from the scariness.

**Comment by I-am-Erk** [comment]
I don't think it's too bad, given that it's a one shot. Masters and necromancers are problems because of the ongoing revivification.

That said, if it's possible I would suggest that it be a *chance* to upgrade, rather than a guarantee, because of the fairly high risk of suddenly facing an unstoppable gang of hulks

**Comment by esotericist** [comment]
what would happen if it just advanced the evolution timer by some number of days?

**Comment by Little-119** [comment]
Sounds !!FUN!!

**Comment by nphyx** [comment]
> I don't like this. Zombie necromancer and zombie master were on the edge with their magic abilities, and this monster steps far over the border. It's just a combined version of two previous monsters with no new abilities, no distinctive features, and, more importantly, no justification. It clearly smells of "difficulty for the sake of difficulty".

Totally disagree with you here.

Right now the only late-game threats from the zombie/blob faction are the tanks, and they're all basically the same thing with different names unless you try to melee them.

Late game desperately needs more tactical and strategic complications. Late-game versions of mid-game threats is a great start. Masters and necros are rarely seen and are trivial to deal with late-game right now (you just snipe them when you see them, and they're so weak it's easy to one-shot them).

This makes the necro/master mechanic relevant again late-game. With one of these things on screen you have to decide between killing it first and dealing with the upgrades vs. hoping you can isolate it later on to avoid the revives (and if you fail the gamble you get the worst of both worlds). This is nice. It's also three mechanics introduced earlier in the game recombined in a novel context, which is nice from a game design perspective.

**Comment by Night-Pryanik** [comment]
Combining several preexisting features into one new monster neither interesting not fun, at least for me. I agree, we need more late-game threats, but not the "combine"-style.
If we have several almost equal monsters, it's a reason to remove some of them, not the reason to use them as precedent. What next - necro-acid-electric-kevlar-feral-skeletal-master-hulk?

**Comment by ScampsAdams** [comment]
Sounds very scary. Even more scary if it's killed accidentally (turrets? followers? traps?)
Lategame butcher shops = chain reaction of necro-boomers?
Necros and masters are limited to one zombie a time, and they are powerful as is. Maybe delay resurrection, say spawn some resurrecting field with some chance/turn?

Another idea: maybe some exploding zombie could spawn a feral or two in its place, alien-style. Cliche, I know.

**Comment by procrustesmethod** [comment]
I think this is a creative way to combine two abilities, actually. I definitely have issues with putting abilities together - an acid runner, or skeletal master, or random variant like that would be bland and wouldn't make a whole lot of sense. 
The weakness of various boomer zombies is, after you have a gas mask or towel or something, they don't pose a threat. The low HP means they won't get any hits on you, and only serve as a early or mid-game challenge. Just ignore it or shoot it once from a distance & problem solved.

necromancers are soft, and although they can be a pain (especially working from a distance) It's pretty easy to pick it off once spotted, then clean up. Even if you've just killed 20 zombies and a necro shows up, it'll only have time to resurrect a few before you kill it.

With the necroboomer, sure you are kind-of just shoving these two together, but it plays much much differently. Unlike any other zombie, you don't want this one dead. You'll have to balance a lot of different factors & change your playstyle as soon as one enters the ring. AoE resurrection upon death is really powerful & is more than its parts (boomer & necromancer) whereas this isn't the case for many other zombies' abilities. 

This is mostly from a gameplay perspective though. As far as lore goes I'm not sure, it might be crossing some line I'm not aware of. But I think it'd be cool to see in-game as an actually challenging late-game boomer.

**Comment by Night-Pryanik** [comment]
> Unlike any other zombie, you don't want this one dead.

That might be the only good thing about this monster. It could potentially suppress the most players' urge to kill _every monster_ they see, while the most "natural" behavior for a survivor should be to evade any encounters with any monster whatsoever.

Also I don't find this monster to pose any challenges. It's doing some dangerous magic on its death (how?)? Ok, then just don't kill it at all, simple. Or, if you _really_ need it to be killed, then kill all other zombies around it first, then lure it somewhere where its abilities won't work, and kill it.

**Comment by KorGgenT** [comment]
> Also I don't find this monster to pose any challenges.

I have to say, I quite disagree with practically everything you're saying about this monster. I would even venture to say that kevlar hulks, shocker hulks, and skeletal juggernauts are the boring monsters, and this one is actually interesting - it could be more interesting if it can decide to explode on its own as well. This is a classic example of using multiple abilities from other monsters in a different way to make something new. It's not just a combination of a zombie master and a zombie necromancer, though it is based on those and I'm not particularly creative with the name. It's using a monster type that I'll even posit we don't have enough of: the kind that explode on death, meaning that killing them doesn't just wipe them off the map, you have to deal with the fallout even if you can kill them before the monster gets close. I haven't seen any new and interesting zombie being proposed recently I have to say, and "there isn't enough endgame" is something that players have been screaming about since the annals of history began. 

The strategy being reversed, not wanting to kill the zombie in a bad situation, is something that will run counter to the murder-hobo strategy that is very common, and really does add something to the game I feel. It can certainly be explained in a lore fashion if you wish, but I don't care as much as the other devs to make up exactly how and why it works - I thought this would be a fun and new zombie mechanically, and that's it.

**Comment by nphyx** [comment]
> Also I don't find this monster to pose any challenges. 

>then kill all other zombies around it first, then lure it somewhere where its abilities won't work, and kill it.

I wonder what you think of as a "challenge". Because you just described the tactical complication this monster introduces perfectly, which is the sort of thing I find to be "challenging".

What I don't find to be challenging is, "need to shoot it 2 more times than the lower tier version of the same monster and 1 more time than the almost identical other monsters on this tier". That sums up the entirety of the "endgame" in this game right now: how many shots of .223 do I need to kill it? Or in other words, how many times do I have to press "f" so I can get on with looting more .223 ammunition?

**Comment by Night-Pryanik** [comment]
> kevlar hulks, shocker hulks, and skeletal juggernauts are the boring monsters

Completely agreed. The "more-armor-more-hp" type of monsters is boring and uninteresting, and honestly I want them removed or at least replaced with something else. Also the monsters you listed are ones that were created by "We have monsters with X and Y features, let's create monster with X+Y feature!" template. I.e. shocker hulk = hulk + shocker. I personally strongly dislike this kind of "creativity", that's why I dislike your "necromancer + master + boomer" monster. But that's only my opinion, if other people find this monster fun and it become added to the game, I'll accept it.

> Because you just described the tactical complication this monster introduces perfectly, which is the sort of thing I find to be "challenging".

"Ignore this monster until it directly threats you" is a tactical complication? "Kill this monster the last" is a tactical complication?

> I wonder what you think of as a "challenge".

The monster you couldn't simply snipe (for various reasons: it's very dodgy, it's invisible, it somehow punishes you for ranged fire, etc) from 50+ tiles I find challenging. The monster you couldn't simply drive over with your deathmobile (for various reasons) I find challenging.

**Comment by nphyx** [comment]
> Completely agreed. The "more-armor-more-hp" type of monsters is boring and uninteresting, and honestly I want them removed or at least replaced with something else.

Well I think we're all on the same page there then. I think here you're making the perfect the enemy of the good.

> "Ignore this monster until it directly threats you" is a tactical complication? "Kill this monster the last" is a tactical complication?

"Isolate this monster and pull it away from the group or corpses" is a complication, yes. It's a pretty classic one as far as tactical RPG gameplay goes. Besides the proactive effort, it also means you are incentivised to avoid using or triggering AoEs, which further complicates the situation. Or would, except why would you bother with AoEs right now, but that's a separate problem.

> The monster you couldn't simply snipe (for various reasons: it's very dodgy, it's invisible, it somehow punishes you for ranged fire, etc) from 50+ tiles I find challenging.

Basically there isn't anything you can't snipe with a big enough gun, and while I'd love to see some monsters that are vulnerable to melee and essentially immune to ranged, we don't have any yet. With the new ballistic damage type I think we have room to make some improvements there.

>  The monster you couldn't simply drive over with your deathmobile (for various reasons) I find challenging.

The only thing that gives me pause with running things over are the ones that create gas effects (boomers, smokers). So that stops being a problem once you and any companions are equipped with gas masks. Bringing back the danger for these kinds of zombies late-game might make vehicular homicide more of a gamble. I'd like that a lot, because I can't see a good reason not to use my vehicle as my main weapon right now.

Unfortunately right now one of those examples is usually the solution to the other (if you can't snipe it run it over; if you can't run it over snipe it).

**Comment by Night-Pryanik** [comment]
> Isolate this monster

It doesn't pose a direct threat to you. It doesn't have any ranged attack, it isn't very quick to seriously outpace you, it can't jump or teleport to quickly shorten the distance between you, it doesn't make its magic spontaneously. In other words, it's a totally mediocre monster until _you_ decide to kill it. There's no need to isolate it when you can just ignore it. 
I'll just use my "everyday" strategy to kill everyone else around it, leaving it for the last moment. At best I can call this an "offshoot" to my usual tactics of dealing with monsters, but I definitely can't call this "challenging". But I repeat, probably that's only me and my opinion.

**Comment by nphyx** [comment]
>it's a totally mediocre monster until _you_ decide to kill it. There's no need to isolate it when you can just ignore it.
That's a fair point. So it should provoke you to kill it for some reason. Or else spontaneously explode after some period of provocation, so that you're inspired to deal with it early on your own terms.

**Comment by KorGgenT** [comment]
I think you guys are forgetting that it is using copy-from, which means it still has the necromancer special attack.

**Comment by Mister-Cultist** [comment]
10/10 I now live in constant fear

**Pull request merged**